include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+       *
+       * @note If the keep flag has been set when start_recording_operations() is
+       * called then the tape is immediately usable after creation, and the
+       * values of the independent variables set by register_dof_values() are
+       * those at which the function is to be evaluated. In this case, a call to
+       * this function is not strictly necessary.
+       */
+      template <typename VectorType>
+      void
+      set_dof_values(
+        const VectorType &                                  values,
+        const std::vector<dealii::types::global_dof_index> &local_dof_indices);
+
+      //@}
+
+      /**
+       * @name Dependent variables
+       */
+      //@{
+
+      /**
+       * Computes the value of the residual vector field
+       * $\mathbf{r}(\mathbf{X})$.
+       *
+       * @return A Vector object with the value for each component of the vector
+       * field evaluated at the point defined by the independent variable
+       * values.
+       */
+      virtual Vector<scalar_type>
+      compute_residual() const = 0;
+
+      /**
+       * Computes the gradient (first derivative) of the residual vector field
+       * with respect to all independent variables, i.e.
+       * @f[
+       *   \frac{\partial\mathbf{r}(\mathbf{X})}{\partial\mathbf{X}}
+       * @f]
+       *
+       * @return A FullMatrix with the gradient of each component of the vector
+       * field evaluated at the point defined by the independent variable
+       * values.
+       */
+      virtual FullMatrix<scalar_type>
+      compute_linearization() const = 0;
+
+      //@}
+
+    }; // class ADHelperCellLevelBase
+
+
   } // namespace AD
 } // namespace Differentiation
 
 
+/* ----------------- inline and template functions ----------------- */
+
+
+#  ifndef DOXYGEN
+
+namespace Differentiation
+{
+  namespace AD
+  {
+    /* ----------------- ADHelperCellLevelBase ----------------- */
+
+
+
+    template <enum AD::NumberTypes ADNumberTypeCode, typename ScalarType>
+    template <typename VectorType>
+    void
+    ADHelperCellLevelBase<ADNumberTypeCode, ScalarType>::register_dof_values(
+      const VectorType &                                  values,
+      const std::vector<dealii::types::global_dof_index> &local_dof_indices)
+    {
+      // This is actually the same thing the set_dof_values() function,
+      // in the sense that we simply populate our array of independent values
+      // with a meaningful number. However, in this case we need to double check
+      // that we're not registering these variables twice
+      Assert(
+        local_dof_indices.size() == this->n_independent_variables(),
+        ExcMessage(
+          "Degree-of-freedom index vector size does not match number of independent variables"));
+#    ifdef DEBUG
+      for (unsigned int i = 0; i < this->n_independent_variables(); ++i)
+        {
+          Assert(this->registered_independent_variable_values[i] == false,
```

Comment:
Just remove the `#ifdef` -- the compiler will optimize away the loop once the `Assert` is expanded to nothing in release mode.